Fix driver message parser

jwillmartin opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description

Changes to the driver have been made to handle cases where the driver might mistake a PSID as the messageID, inadvertently sending the wrong payload to CARMA Messenger to be decoded. The driver now checks for messageIDs and PSIDs along with expected message length to parse for the correct payload. 

Note: An additional change was made to the PSM PSID to correct the value.

## Related GitHub Issue

<!--- This project only accepts pull requests related to open GitHub issues or Jira Keys -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please DO NOT name partially fixed issues, instead open an issue specific to this fix -->
<!--- Please link to the issue here: -->

## Related Jira Key

VS-25

## Motivation and Context

Change was required to correctly parse for the payload received from an OBU. Most OBUs do not have an application to forward V2X payloads, so a generic packet forwarding tool (netcat or socat) is used to forward all radio packets back to CARMA when needed. Forwarding all packets means there is a need to parse for the actual V2X payload. These updates are able to parse for the payload when an entire radio packet is received, along with parsing for the V2X payload when it is sent in an ideal format. Additionally, any non-V2X messages will be ignored. 

This issue was first found by OSU when they used an OBU in a way CARMA does not. This caused incorrect messages to be sent to CARMA Messenger. These fixes address their issues. 

## How Has This Been Tested?

Tested locally with messages created by CARMA and with messages captured from OSU, who first reported the issue. 

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [X] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

jwillmartin committed:
new file:   .devcontainer/Dockerfile.builder
	new file:   .devcontainer/devcontainer.json
	new file:   .devcontainer/docker-compose-vscode.yml
	new file:   .vscode/tasks.json
	modified:   Dockerfile
	modified:   docker/install.sh
	modified:   dsrc_driver/src/dsrc_client.cpp

jwillmartin committed:
modified:   dsrc_driver/include/dsrc_driver/dsrc_client.h
	modified:   dsrc_driver/src/dsrc_client.cpp

jwillmartin committed:
modified:   .vscode/tasks.json

jwillmartin committed:
modified:   .vscode/tasks.json
	modified:   Dockerfile
	modified:   docker/install.sh

jwillmartin committed:
Merge remote-tracking branch 'origin/develop' into fix-driver

jwillmartin committed:
modified:   docker/install.sh
	modified:   dsrc_driver/include/dsrc_driver/dsrc_client.h
	modified:   dsrc_driver/src/dsrc_client.cpp

jwillmartin committed:
Merge branch 'fix-driver' of https://github.com/usdot-fhwa-stol/carma-cohda-dsrc-driver into fix-driver

jwillmartin committed:
modified:   dsrc_driver/include/dsrc_driver/dsrc_client.h
	modified:   dsrc_driver/src/dsrc_client.cpp

jwillmartin committed:
modified:   dsrc_driver/config/wave.json
	modified:   dsrc_driver/etc/wave.json
	modified:   dsrc_driver/include/dsrc_driver/dsrc_client.h
	modified:   dsrc_driver/launch/dsrc_driver.py
	modified:   dsrc_driver/src/dsrc_client.cpp

afrye51 committed:
Merge branch 'develop' into fix-driver

afrye51 committed:
Fixed driver_shutdown_ros2 not being build

jwillmartin committed:
Merge remote-tracking branch 'origin/develop' into fix-driver

jwillmartin committed:
Merge remote-tracking branch 'origin/develop' into fix-driver

jwillmartin committed:
modified:   v2x_ros_driver/src/v2x_radio_client.cpp

afrye51 commented on v2x_ros_driver/config/wave.json:
```diff
@@ -64,7 +64,7 @@
   {
     "name" : "PSM",
-    "psid" : "0082",
+    "psid" : "0027",
```
What are these changes exactly?

jwillmartin replied:
This is a fix for the PSID. Correct is 0x27 from this site:
https://standards.ieee.org/products-programs/regauth/psid/public/

afrye51 commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -174,23 +175,131 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
-            break;
+
+            // Check if specified MessageFrame size matches actual data size
+            printVector(msg_vec);
```
Could you remove these prints (also line 197) or make them trigger on log level? I assume debug

jwillmartin replied:
Added ros2 info logs

JonSmet replied:
I think we may want to write to ros2 debug stream, if we write to ros2 info stream, the terminal will fill with logs for incoming messages. Thoughts?

jwillmartin replied:
Good idea, I agree. Updated to a debug stream

afrye51 commented on v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h:
```diff
@@ -134,13 +154,18 @@ class V2XRadioClient
     boost::asio::ip::udp::endpoint remote_udp_ep_;
     std::unique_ptr<cav::UDPListener> udp_listener_;
 
+    int long_frame = 4;
```
Could you make these three new variables static, with an _ at the end (long_frame_)

JonSmet replied:
Can you also add a brief comment for each of new variables?

jwillmartin replied:
Made static, added briefs

afrye51 commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -174,23 +175,131 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+            }
+        else 
+        {
+            std::cerr << "Size in possible MessageFrame does not match actual data size. Checking rest of data." << std::endl;
```
use ros error stream

jwillmartin replied:
added ros2 warn streams

afrye51 commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -174,23 +175,131 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+                return true;
+            }
+        else {
+            std::cerr << "Size in possible MessageFrame does not match actual data size. Checking rest of data." << std::endl;
```
use ros error stream

jwillmartin replied:
added ros2 warn streams

afrye51 commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -174,23 +175,131 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+
+void V2XRadioClient::printVector(const std::vector<uint8_t>& vec)
+{
+    std::cerr << "[";
```
use ros info/debug stream

jwillmartin replied:
added ros2 info stream

JonSmet commented on .devcontainer/docker-compose-vscode.yml:
```diff
@@ -0,0 +1,78 @@
+#  Copyright (C) 2018-2023 LEIDOS.
```
CARMA Platform already had a [documented process](https://usdot-carma.atlassian.net/wiki/spaces/CRMPLT/pages/488505352/Setup+VS+Code+for+CARMA+Platform+Development) for using a devcontainer for development with its related repositories. Since v2x-ros-driver can be used outside of CARMA Platform, maybe it is useful to maintain its own devcontainer, but can we remove or revise this docker compose since it is requiring ros1_bridge and carma-messenger services? 

Might be worth setting up a quick 30 minute meeting with @MishkaMN and @adev4a to briefly discuss devcontainer support for our auxiliary CARMA Platform repos so that we can have a consistent approach that we can reuse in all of our repos. If it spins off more questions, maybe we hold these changes off for a separate PR/task?

jwillmartin replied:
Branched to feature-dev-container, removed dev container and vscode task from this branch.

JonSmet commented on Dockerfile:
```diff
@@ -27,7 +29,8 @@ COPY --chown=carma . /home/carma/src/
 
-FROM setup
+# Final image
+FROM setup AS final
```
Is this a change to match best practices for Dockerfiles?

jwillmartin replied:
Yeah, it's a change to match best practice. Docker builder also complained about it when building

JonSmet commented on v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h:
```diff
@@ -123,7 +123,27 @@ class V2XRadioClient
+    /**
+     * @brief Checks actual message vector size against size specified in MessageFrame
+     */
+    bool isValidMsgSize(const std::vector<uint8_t> msg_vec, size_t start_index, size_t end_index, const std::vector<uint8_t> entry);
```
While this file was made by an external company and does not contain additional comments for functions, would you mind adding comments for the function arguments as well for these new functions? Similar to [this example in basic_autonomy](https://github.com/usdot-fhwa-stol/carma-platform/blob/7888a6e80c16f55383866e9102085433c8d490ed/basic_autonomy/include/basic_autonomy/basic_autonomy.hpp#L122)

jwillmartin replied:
Added param descriptions

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -174,23 +175,131 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+        if (msg_id == psidInt)
+            return true;
+    }
+    return false;
```
Can we add a log statement to ROS ERROR Stream here as well?

jwillmartin replied:
Added ros2 warn stream

JonSmet commented on Dockerfile:
```diff
@@ -11,10 +11,12 @@
 #  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 #  License for the specific language governing permissions and limitations under
 #  the License.
+
```
Placing this comment here but it is more generic to the PR - could you please add a bit more detail regarding the need for this fix (OSU's setup has the OBU generate BSMs, which are forwarded to the driver, which is different from the typical setup used for CARMA Platform testing). Additionally, can we change the PR name to more closely reflect the fix that has been implemented?

Finally, for other changes/updates in the PR that are not relevant to the primary fix, could you please include a note in the PR description for them? I believe these are [1] the new devcontainer support (if they remain in the PR) and [2] the PSM psid fix.

jwillmartin replied:
@JonSmet Updated!

jwillmartin committed:
modified:   v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h
	modified:   v2x_ros_driver/include/v2x_ros_driver/v2x_ros_driver_node.h
	modified:   v2x_ros_driver/src/v2x_radio_client.cpp

jwillmartin committed:
modified:   v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h
	modified:   v2x_ros_driver/src/v2x_radio_client.cpp

jwillmartin committed:
deleted:    .devcontainer/Dockerfile.builder
	deleted:    .devcontainer/devcontainer.json
	deleted:    .devcontainer/docker-compose-vscode.yml
	deleted:    .vscode/tasks.json

JonSmet commented on v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h:
```diff
@@ -106,49 +104,85 @@ class V2XRadioClient
+    bool isValidMsgSize(const std::vector<uint8_t> msg_vec, size_t start_index, size_t end_index, const std::vector<uint8_t> entry);
+    
+    /**
+     * @brief Checks msg_id against PSID list to make sure a WSA with valid PSID is not accidentally used instead of MessageFrame
```
What is "WSA" in this context? Might be worth spelling it out

jwillmartin replied:
Updated

JonSmet commented on v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h:
```diff
@@ -106,49 +104,85 @@ class V2XRadioClient
+     * @brief Helper function to print vectors for debugging
+     * @param vec Vector to be printer
+     */
+    void printVector(const std::vector<uint8_t>& vec);
```
Can we adjust this function name and description to indicate that this is a helper function for outputting the vectors to a ROS Stream (looks like we're still determining whether an info or error stream is needed), rather than printing them?

jwillmartin replied:
Updated

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,138 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
             // TODO lengths greater than 16383 (0x3FFF) are encoded by splitting up the message into discrete chunks, each with its own length
             // marker. It doesn't look like we'll be receiving anything that long
-            std::cerr << "V2XRadioClient::process() : received a message with length field longer than 16383." << std::endl;
+            RCLCPP_WARN_STREAM(logger_, "V2XRadioClient::process() : received a message with length field longer than 16383.");
```
I know the log text is unchanged, but can we add that the message is being discarded?

jwillmartin replied:
Updated

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,138 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+            {
+                if (!isValidPSID(std::to_string(msg_id)))
+                {
+                    printVector(msg_vec);
```
For this valid message, we've now called printVector() twice, which will result in excessive logging. Is printVector() needed at all in this function, since [onMessageReceivedHandler()](https://github.com/usdot-fhwa-stol/v2x-ros-driver/blob/58cc0cf3c0ee0ead9ecd7e1b845f8799a57d4b56/v2x_ros_driver/src/v2x_ros_driver_node.cpp#L155) already logs to RCLCPP_DEBUG_STREAM()?

jwillmartin replied:
Good catch, accidentally left there while debugging. Removed first function call.

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,138 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+
+            // Check if specified MessageFrame size matches actual data size
+            printVector(msg_vec);
+            if (!isValidMsgSize(msg_vec, start_index, end_index, entry))
```
It is is a little confusing to step through these nested if-else statements, could a brief comment be added to the branches to help a new user looking at this (i.e. why is a message with a valid message size and an invalid PSID considered a valid message?)

jwillmartin replied:
Updated

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,138 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+    RCLCPP_INFO_STREAM(logger_, "Sending message vector: " << oss.str());
+}
+
+std::vector<uint8_t> V2XRadioClient::hexStringToByteArray(const std::string& hexString) const
```
It looks like this function is no longer used, can we remove it?

jwillmartin replied:
Removed

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,138 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+            }
+        else 
+        {
+            RCLCPP_WARN_STREAM(logger_, "Size in possible MessageFrame does not match actual data size. Checking rest of data.");
```
I think we need to move all or part of this log to the calling function. This log states "Checking rest of data." but `isValidMsgSize()` does not do anymore checks, they're handled in `process()`

jwillmartin replied:
Updated

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,138 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+                return true;
+            }
+        else {
+            RCLCPP_WARN_STREAM(logger_, "Size in possible MessageFrame does not match actual data size. Checking rest of data.");
```
I think we need to move all or part of this log to the calling function. This log states "Checking rest of data." but `isValidMsgSize()` does not do anymore checks, they're handled in `process()`

jwillmartin replied:
Updated

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,138 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+        // Convert the integer to its decimal string representation
+        std::string psidInt = std::to_string(psid_value);
+        if (msg_id == psidInt)
+            RCLCPP_WARN_STREAM(logger_, "PSID found, parsing rest of data for MessageID.");
```
I think we need to move all or part of this log to the calling function. This log states "parsing rest of data..." but `isValidPSID()` does not do anymore checks, they're handled in `process()`

jwillmartin replied:
Good call, updated

jwillmartin committed:
modified:   v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h
	modified:   v2x_ros_driver/src/v2x_radio_client.cpp

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,124 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+                // Make sure WSA is not accidentally detected before actual message. This is done by checking detected msg_id against list of PSIDs.
+                if (!isValidPSID(std::to_string(msg_id)))
+                {
+                    printVectorHelper(msg_vec);
```
Do you think `printVectorHelper()` is still needed since `OnMessageReceivedHandler()` (which is called by `OnMessageReceived()`) already prints the vector to RCLCPP_DEBUG_STREAM ([see here](https://github.com/usdot-fhwa-stol/v2x-ros-driver/blob/58cc0cf3c0ee0ead9ecd7e1b845f8799a57d4b56/v2x_ros_driver/src/v2x_ros_driver_node.cpp#L155C115-L155C141)). Or do `printVectorHelper()` and `uint8_vector_to_hex_string()` format the vector differently?

jwillmartin replied:
They would be in the same format, but printVectorHelper would will only print out the valid message vs OnMessageReceived printing out all received messages. Think we should leave this in?

jwillmartin replied:
Removed helper function

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,124 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+    {
+        auto tmp_start_index = start_index + short_frame_;
+        std::vector<uint8_t> short_vec(entry.begin() + tmp_start_index, entry.begin() + end_index);
+        auto msg_size = msg_vec[2];
```
Can we add a check first to make sure msg_vec.size() is large enough to access this element to avoid an out of range exception?

jwillmartin replied:
Updated to make sure msg_vec[2] exists

jwillmartin committed:
modified:   v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h
	modified:   v2x_ros_driver/src/v2x_radio_client.cpp

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -168,29 +169,110 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+        }
+    }
+    // If message vector is smaller than 255 bytes, length field will be 1 octet. Additional check to make sure msg_size[2] exists.
+    else if (msg_vec.size() < 255 && msg_vec.size() > 3)
```
This needs to be updated to `... msg_vec.size() >= 3`

jwillmartin committed:
modified:   v2x_ros_driver/src/v2x_radio_client.cpp

afrye51 reviewed (approved):
Tested in simulation
